### DISCUSSION TITLE
Handling broken connections with the server

### DISCUSSION THREAD

**Discussion by MaciejPMarciniak** [FreeOpcUa/opcua-asyncio#1665] [Q&A]
Hi, I am facing the following issue: the OPCUA server that I am working with is a spoiled brat. 

Within an hour of beginning the subscription, I get the error about the broken connection from the watchdog loop:
```
Traceback (most recent call last):
  File "/usr/local/lib/python3.12/asyncio/tasks.py", line 520, in wait_for
    return await fut
           ^^^^^^^^^
asyncio.exceptions.CancelledError

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/usr/local/lib/python3.12/site-packages/asyncua/client/client.py", line 551, in _monitor_server_loop
    _ = await self.nodes.server_state.read_value()
        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.12/site-packages/asyncua/common/node.py", line 207, in read_value
    result = await self.read_data_value()
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.12/site-packages/asyncua/common/node.py", line 220, in read_data_value
    return await self.read_attribute(ua.AttributeIds.Value, None, raise_on_bad_status)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.12/site-packages/asyncua/common/node.py", line 342, in read_attribute
    result = await self.session.read(params)
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.12/site-packages/asyncua/client/ua_client.py", line 404, in read
    data = await self.protocol.send_request(request)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.12/site-packages/asyncua/client/ua_client.py", line 167, in send_request
    data = await asyncio.wait_for(self._send_request(request, timeout, message_type), timeout if timeout else None)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.12/asyncio/tasks.py", line 519, in wait_for
    async with timeouts.timeout(timeout):
  File "/usr/local/lib/python3.12/asyncio/timeouts.py", line 115, in __aexit__
    raise TimeoutError from exc_val
```

I saw similar issues in the discussions, and the example with reconnection. However, to add to the pile, I only have **ONE** license to connect to the server, which is why attempting to reconnect causes the **BadLicenseLimitExceeded** error. The only way to resubscribe is to **restart the server** and connect again.

Does anyone experienced with OPCUA know:

1. Why the servers may be losing connection? I run the same code on other servers and they are stable for days (weeks).
2. Is there a way to reconnect on the same license, without the need for a reboot?

The troublesome server is GE iFIX 6.0.

Thanks and I hope I am asking in the correct format!

**Comment by oroulet** [comment]
sounds like the tcp server died. It can be a bug in the server or simply bad network. you could try to monitor the connection using wireshark .it has very good support for U, and see if the server send some more information that we do not display by mistake

**Reply by MaciejPMarciniak** [reply]
I've checked the wireshark and it seems like nothing suspicious is happening, except suddenly the server stops responding, after a random amount of time:

![image](https://github.com/FreeOpcUa/opcua-asyncio/assets/45173133/f60fdc55-cf67-4322-bf4e-b647fbc43a1a)

Afterwards no more packets are sent to the destination address 172.28.34.135, which marks the opcua server address. Does it mean that it's a server bug?

**Comment by AnushkaThawani** [comment]
Were you able to resolve this issue in the end? I'm facing something similar
